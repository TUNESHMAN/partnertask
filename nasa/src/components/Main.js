import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { Spin, Button, DatePicker } from "antd";
import { fetchPhoto } from "../state/actions/photoAction";
import * as moment from "moment";
import "../App.css";
import { NavLink } from "react-router-dom";
import SuccessModal from "./SuccessModal";

function Main(props) {
  const dateFormat = "YYYY/MM/DD";
  var d = new Date();
  const today = new Date(d.setDate(d.getDate())).toLocaleDateString("fr-CA");
  const [chosen_Date, setChosen_Date] = useState(today);

  const handleDate = (value) => {
    setChosen_Date(value._d.toLocaleDateString("fr-CA"));
    props.fetchPhoto(chosen_Date);
  };

  // This function controls the previous button
  const previous = (date) => {
    var d = new Date(chosen_Date);
    const yesterday = new Date(d.setDate(d.getDate() - 1)).toLocaleDateString(
      "fr-CA"
    );
    setChosen_Date(yesterday);
    props.fetchPhoto(yesterday);
  };

  // This function controls the next button
  const next = (date) => {
    var d = new Date(chosen_Date);
    const tomorrow = new Date(d.setDate(d.getDate() + 1)).toLocaleDateString(
      "fr-CA"
    );
    setChosen_Date(tomorrow);
    props.fetchPhoto(tomorrow);
  };

  // I use the useEffect hook to carry out component side effect
  useEffect(() => {
    props.fetchPhoto(chosen_Date);
  }, [chosen_Date]);

  // Favorites
  const [favorites, setFavorites] = useState([]);

  const handleFavorites = (e, fav) => {
    setFavorites((prev) => [...prev, fav]);
    localStorage.setItem("favoriteList", JSON.stringify(favorites));
    for (let i = 0; i < favorites.length; i++) {
      const greaterIndex = favorites[i++];
      let count = 0;
      if (greaterIndex) {
        count++;
        SuccessModal();
      }
    }
  };

  return (
    <div className="body">
      {props.isFetching ? (
        <div className="spinner">
          <Spin size="large" spinning={props.isFetching} />
        </div>
      ) : Object.keys(props.photo).length === 0 ? (
        <div className="error-div">
          <h2 className="error-text">No photo to display</h2>
        </div>
      ) : (
        <div>
          <div className="container">
            <h1>Photo of the day for {props.photo.date}</h1>
            <header>
              <h2>
                <span>Title:</span> {props.photo.title}
              </h2>
            </header>

            <div className="image-div">
              <img
                src={props.photo.hdurl}
                alt="photo of the day"
                style={{ width: "100%" }}
              />
              <div className="description-holder">
                <p>{props.photo.explanation}</p>
              </div>
            </div>
            <div className="btn">
              <Button className="btn-btn" onClick={previous}>
                Previous
              </Button>
              <Button className="btn-btn" onClick={next}>
                Next
              </Button>
              <DatePicker
                name="chosen_Date"
                setFieldsValue={moment(chosen_Date, dateFormat)}
                format={dateFormat}
                onChange={handleDate}
                placeholder="Date"
                className="btn-btn"
              />
              <Button
                className="btn-btn1"
                onClick={(e) => handleFavorites(e, props.photo)}
              >
                Set as Favorite
              </Button>
              <NavLink to="/favorites">
                <Button className="btn-btn1">View Favorite</Button>
              </NavLink>
            </div>
          </div>
        </div>
      )}
    </div>
  );
}

// mapStateToProps function is the connection between our state from redux and the main component
const mapStateToProps = (state) => {
  return {
    // I return an object with each property from the initial state declared in the reducer
    isFetching: state.photo.isFetching,
    error: state.photo.error,
    photo: state.photo.picture,
  };
};
// I wrap the export of the main component inside the connect call
export default connect(mapStateToProps, { fetchPhoto })(Main);
