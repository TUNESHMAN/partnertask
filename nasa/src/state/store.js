// Thunk is a middleware
import thunk from "redux-thunk";
import { combineReducers, createStore, compose, applyMiddleware } from "redux";
import photoReducer from "./reducers/photoReducer";

// This houses the other reducers
const mainReducer = combineReducers({ photo: photoReducer });

const devTools =
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
    ? window.__REDUX_DEVTOOLS_EXTENSION__ &&
      window.__REDUX_DEVTOOLS_EXTENSION__()
    : (n) => n;

const store = createStore(
  mainReducer,
  {},
  compose(applyMiddleware(thunk), devTools)
);

export default store;
