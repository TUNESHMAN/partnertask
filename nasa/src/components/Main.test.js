import React from "React";
import Main from "./Main";
import * as rtl from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Provider } from "react-redux";
import store from "../state/store";

// Take care of cleaning up after every operation
afterEach(rtl.cleanup);

// Handle repetitive operations inside a beforeEach function
let wrapper;
beforeEach(() => {
  // The aim is to recreate the "wrapper" at every test
  wrapper = rtl.render(
    <Provider store={store}>
      <Main />
    </Provider>
  );
});
// Make a trivial test
it("renders without crashing", () => {
  wrapper.debug(); //This is like console.log. It gives a look into the fake Dom that is created
  expect(wrapper.container).toMatchSnapshot();
});
