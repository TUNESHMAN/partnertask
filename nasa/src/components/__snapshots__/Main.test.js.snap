// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <div
    class="body"
    data-testid="container"
  >
    <div>
      <div
        class="container"
      >
        <h1>
          Photo of the day for 
        </h1>
        <header>
          <h2>
            <span>
              Title:
            </span>
             
          </h2>
        </header>
        <div
          class="image-div"
        >
          <img
            alt="photo of the day"
            style="width: 100%;"
          />
          <div
            class="description-holder"
          >
            <p />
          </div>
        </div>
        <div
          class="btn"
        >
          <button
            class="ant-btn btn-btn"
            data-testid="previous"
            type="button"
          >
            <span>
              Previous
            </span>
          </button>
          <button
            class="ant-btn btn-btn"
            data-testid="next"
            type="button"
          >
            <span>
              Next
            </span>
          </button>
          <span
            class="btn-btn ant-calendar-picker"
          >
            <div>
              <input
                class="ant-calendar-picker-input ant-input"
                name="chosen_Date"
                placeholder="Date"
                readonly=""
                value=""
              />
              <i
                aria-label="icon: calendar"
                class="anticon anticon-calendar ant-calendar-picker-icon"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="calendar"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                  />
                </svg>
              </i>
            </div>
          </span>
          <button
            class="ant-btn btn-btn1"
            type="button"
          >
            <span>
              Set as Favorite
            </span>
          </button>
          <a
            href="/favorites"
          >
            <button
              class="ant-btn btn-btn1"
              data-testid="viewFav"
              type="button"
            >
              <span>
                View Favorite
              </span>
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
